scripts/thumbnail-generator/generate_thumbnail.py

```diff
@@ -0,0 +1,302 @@
+    # Focal length and sensor size should be in the same units (e.g., mm)
+    field_of_view = 2 * math.atan(sensor_size / (2 * focal_length))
+    return field_of_view
+
+def calculate_camera_distance(subject_size, field_of_view):
+    # Subject size and field of view should be in the same units (e.g., mm and degrees)
+    distance = (subject_size / 2) / math.tan(field_of_view / 2)
+    return distance
+
+def get_camera_z_translation(distance):
+    return Gf.Vec3d(0, 0, distance)
+
+def apply_camera_translation(camera_stage, camera_prim, camera_translation):
+    xformRoot = UsdGeom.Xformable(camera_prim.GetPrim())
+    translateOp = None
+    # Go through each operation in the xformable schema
+    for op in xformRoot.GetOrderedXformOps():
+        # If the operation is a translate operation, we've found our operation
+        if op.GetOpType() == UsdGeom.XformOp.TypeTranslate:
+            translateOp = op
+            break
+
+    # If no translate operation exists, create one
+    if translateOp is None:
+        translateOp = xformRoot.AddTranslateOp(UsdGeom.XformOp.PrecisionDouble)
+
+    translateOp.Set(camera_translation)
+    camera_stage.Save()
+
+def sublayer_subject(camera_stage, input_file):
+    camera_stage.GetRootLayer().subLayerPaths = [input_file]
+    camera_stage.GetRootLayer().Save()
+
+def take_snapshot(image_name):
+    renderer = get_renderer()
+    cmd = ['usdrecord', '--camera', 'MainCamera', '--imageWidth', str(args.width), '--renderer', renderer, 'camera.usda', image_name]
+    run_os_specific_usdrecord(cmd)
+    os.remove("camera.usda")
+    if os.path.isfile("y_up.usda"):
+        os.remove("y_up.usda")
+    return image_name
+
+def get_renderer():
+    if os.name == 'nt':
+        print("windows default renderer GL being used...")
+        return "GL"
+    else:
+        if sys.platform == 'darwin':
+            print("macOS default renderer Metal being used...")
+            return 'Metal'
+        else:
+            print("linux default renderer GL being used...")
+            return 'GL'
+
+def run_os_specific_usdrecord(cmd):
+    if os.name == 'nt':
+        subprocess.run(cmd, check=True, shell=True, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+    else:
+        if sys.platform == 'darwin':
+            subprocess.run(cmd, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+        else:
+            subprocess.run(cmd, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.STDOUT)
+
+def create_image_filename(input_path, extension):
+    return Path(input_path).with_suffix("." + extension)
+
+def link_image_to_subject(subject_stage, image_name):
+    subject_root_prim = subject_stage.GetDefaultPrim()
+    mediaAPI = UsdMedia.AssetPreviewsAPI.Apply(subject_root_prim)
+    thumbnails = UsdMedia.AssetPreviewsAPI.Thumbnails(defaultImage = Sdf.AssetPath(image_name))
+    mediaAPI.SetDefaultThumbnails(thumbnails)
+    subject_stage.GetRootLayer().Save()
```

Comment:
I don't recall the outcome of the discussion on this point in our meeting on Nov 2, but re-saving the input layer seems suspect to me, even though I understand the urge to properly associate the thumbnail with the asset. Maybe make the running of this code optional based on a command line argument?

Reply:
Makes sense, yeah the only reason for this is to associate the image as the asset's thumbnail, will update